src/components/LocaleContextProvider.tsx

```diff
@@ -0,0 +1,149 @@
+import React, {createContext, useMemo} from 'react';
+import {withOnyx} from 'react-native-onyx';
+import {ValueOf} from 'type-fest';
+import ONYXKEYS from '../ONYXKEYS';
+import * as Localize from '../libs/Localize';
+import DateUtils from '../libs/DateUtils';
+import * as NumberFormatUtils from '../libs/NumberFormatUtils';
+import * as LocaleDigitUtils from '../libs/LocaleDigitUtils';
+import CONST from '../CONST';
+import TIMEZONES from '../TIMEZONES';
+import compose from '../libs/compose';
+import withCurrentUserPersonalDetails from './withCurrentUserPersonalDetails';
+import * as LocalePhoneNumber from '../libs/LocalePhoneNumber';
+
+type SelectedTimezone = {
+    /** Value of the selected timezone */
+    selected: (typeof TIMEZONES)[number];
+};
+
+type CurrentUserPersonalDetails = {
+    /** Timezone of the current user */
+    timezone?: SelectedTimezone;
+};
+
+type LocaleContextProviderOnyxProps = {
+    /** The user's preferred locale e.g. 'en', 'es-ES' */
+    preferredLocale: ValueOf<typeof CONST.LOCALES>;
+};
+
+type LocaleContextProviderProps = LocaleContextProviderOnyxProps & {
+    /** The current user's personalDetails */
+    currentUserPersonalDetails: CurrentUserPersonalDetails;
+
+    /** Actual content wrapped by this component */
+    children: React.ReactNode;
+};
+
+type Translate = (phrase: string, variables: Record<string, string>) => string;
```

Comment:
@MaciejSWM let's make this type smarter, here is my idea:

```suggestion
type Translate = <TKey extends TranslationPaths>(
    phrase: TKey,
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    variables?: TranslateType<EnglishTranslation, TKey> extends (...args: any) => any ? Parameters<TranslateType<EnglishTranslation, TKey>>[0] : never,
) => string;
```

This type will allow us to only use the keys that exist and, when is a key that needs variables, we will have type inference too.

```ts
translate('addDebitCardPage.nameOnCard');

translate('chronos.oooEventSummaryFullDay', {
    summary: '',
    dayCount: 1,
    date: '',
});
```

Note: You will have to export `TranslateType` and `TranslationPaths` types from `src/languages/types.ts`.

cc @blazejkustra

Reply:
@fabioh8010 We implemented a similar approach in this [PR](https://github.com/Expensify/App/pull/29742/files) (src/libs/Localize/index.ts, `translate` function). FYI Bartosz is still working on it

Reply:
@MaciejSWM @blazejkustra The way it was implemented in [that PR](https://github.com/Expensify/App/pull/29742/files) works better, let's use it then!